### PR TITLE
Add TypeScript types

### LINKED ISSUE #1: TypeScript types
Would be nice if this library would provide TS types as well. I'm using the following WIP types in my app locally right now:

```ts
// TODO remove this once `@sqlite.org/sqlite-wasm` has types
declare module '@sqlite.org/sqlite-wasm' {
  export type TODO = any

  interface Window {
    // eslint-disable-next-line @typescript-eslint/consistent-type-imports
    sqlite3InitModule: (_?: TODO) => Promise<import('./sqlite3').Sqlite3>
  }

  export type Sqlite3 = {
    oo1: {
      DB: new (options?: OO1DBOptions) => DB
      OpfsDb: new (filename: string, mode: string) => DB
    }
    opfs?: {
      debug: TODO
      deleteEntry: TODO
      entryExists: TODO
      metrics: TODO
      mkdir: TODO
      randomFilename: TODO
      registerVfs: TODO
      rootDirectory: TODO
    }
    capi: {
      sqlite3_deserialize: TODO
      [key: string]: TODO
    }
    wasm: {
      allocFromTypedArray: (typedArray: Uint8Array) => number
      [key: string]: TODO
    }
  }

  /** https://sqlite.org/wasm/doc/trunk/api-oo1.md#db-ctor */
  export type OO1DBOptions = {
    /** @default ':memory:' */
    filename?: string
    /** open-mode flags */
    flags?: string
    /** name of the sqlite3_vfs to use */
    vfs?: string
  }

  export declare const InitWasm: () => Promise<Sqlite3>

  export default InitWasm

  export class DB {
    filename: string
    pointer: TODO

    affirmOpen(): DB

    changes(total: boolean, sixtyFour: boolean): number

    exec(options: ExecOptions): TODO
    exec(query: FlexibleString, options?: ExecOptions): TODO
    close(): DB
    export(): TODO
    /**
     * @example aDb.prepare("INSERT INTO foo(a) VALUES(?)").bind(123).stepFinalize();
     */
    prepare(query: FlexibleString): Stmt

    checkRc(db: TODO, resultCode: TODO)

    createFunction(): TODO

    dbFilename(): string

    dbName(): string
  }

  export type ExecOptions = {
    sql?: FlexibleString
    bind?: Bindable
    // saveSql?: TODO

    returnValue?: 'this' | 'resultRows' | 'saveSql'

    rowMode?: 'array' | 'object' | 'stmt'

    resultRows?: TODO[]
  }

  export type FlexibleString = string | Uint8Array | Int8Array | string[]

  /** https://sqlite.org/wasm/doc/trunk/api-oo1.md#stmt-properties */
  export interface Stmt {
    columnCount: number
    parameterCount: number
    pointer: TODO

    bind(bindable: Bindable): Stmt
    bind(indexOrParameterName: number | string, bindable: Bindable): Stmt

    stepFinalize(): boolean
  }

  export type Bindable = BindableValue[] | Record<string, BindableValue>

  export type BindableValue = null | undefined | number | boolean | string | Uint8Array | Int8Array
}

```

### PR DESCRIPTION
**Pull request by tomayac**
Fixes #1

### CODE REVIEW

**Comment by delaneyj** [comment]
https://gist.github.com/mizchi/cb572eae55154ec781ced5c111621939 seems to be far more complete to build off of

**Comment by lrowe** [comment]
Experimenting with this by adding `/// <reference types="./index.d.ts" />` before my import I found I needed to make the following changes to clear the typescript errors I was seeing from deno:

* Remove all the declare keywords except for the `declare module`. See: https://stackoverflow.com/a/17918684
* Add `interface JsStorageDb extends DatabaseApi {}` before the `export class JsStorageDb implements DatabaseApi`. Same for `class OpfsDatabase` and `class Database`. See: https://stackoverflow.com/a/65787814
* Make the sqlite3InitModule argument optional to match the documentation example at https://sqlite.org/wasm/doc/trunk/demo-123.md:
```ts
  export default function sqlite3InitModule(
    opts?: InitOptions
  ): Promise<Sqlite3Static>;
```

**Comment by tomayac** [comment]
Is this ready for merging, TypeScript people?

**Comment by lrowe** [comment]
I still see a typescript error reported when I call `sqlite3InitModule()` without arguments as the example from the documentation.

**Comment by lrowe** [comment]
Where should these types ultimately live? From https://github.com/tomayac/sqlite-wasm/pull/6#issuecomment-1529956910:

> Another objective is to absolutely limit the number of tools that it depends on itself. In the midterm, the code is to be donated to the [SQLite development team](https://github.com/sqlite/) to allow them to [deploy new versions to npm](https://github.com/tomayac/sqlite-wasm#deploying-a-new-version) easily.

Given that objective and SQLite being [Open-Source, not Open-Contribution](https://www.sqlite.org/copyright.html), would it make more sense to explicitly place them under community maintenance as part of [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)?

**Comment by tomayac** [comment]
If DefinitelyTyped are fine with hosting the code, this would probably be the more future-proof venue. Else, I'm happy to have the TS types here as well. But since the SQLite team are not really into the TS ecosystem, DefinitelyTyped seems indeed a lot more suited. We could still collaborate on getting a PR in shape here, and then migrate it over.

**Comment by sgbeal** [comment]
FWIW, from the perspective of the sqlite team: we're more than happy to have any non-vanilla-JS builds hosted by whoever is willing and able to do so. We don't have the developer bandwidth to keep up with Everyone's Favorite Toolchain and aim to provide the lower-level pieces which can be used to plug sqlite in to those tools by people well-versed in them.

i am happy to assist upstream however i can, e.g. making any code changes which simplify such integration.

**Comment by patdx** [comment]
Looks okay to me as a regular TS user. Though I would suggest to drop the `declare module '@sqlite.org/sqlite-wasm'` wrapper and use top level exports/declarations instead. The "declare module" wrapper makes it painful to alias a a package, for example, if if am using multiple versions in one project and I alias the package as "sqlite-3.42.0"

```js
{
  "name": "sqlite-test",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "sqlite-3.42.0": "npm:@sqlite.org/sqlite-wasm@^3.42.0-build1",
  }
}
```

The type declarations would not resolve when I do `import sqlite from 'sqlite-3.42.0'` in my code, because the `@sqlite.org/sqlite-wasm` is hardcoded in.

It worked fine in my testing to not have the `declare module ...` wrapper. [I changed a few bare classes to "declare class"](https://gist.github.com/patdx/2ccbcc722e280105997f810009b3b6c6) instead which seemed to resolve some minor TypeScript error.

The "unwrapped" style can be found in both integrated packages and DefinitelyTyped declarations so I think you can still decide whether to integrate the types or put on DefinitelyTyped later:

https://unpkg.com/browse/vue@3.3.4/dist/vue.d.ts
https://unpkg.com/browse/@types/three@0.152.1/index.d.ts

**Comment by tomayac** [comment]
Thank you! If others agree (👍) with @patdx‘s assessment, please go ahead and make these modifications.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+      xCursor: any;
+      xIndexInfo: any;
+      xError: any;
+      xRowid: any;
+      setupModule: any;
+    };
+  };
+
+  export default function sqlite3InitModule(
+    opts: InitOptions,
```
Comment (kdabir) [inline index.d.ts:183]: shouldn't it be optional? `opts?`

Comment (tomayac) [inline index.d.ts:183]:
```suggestion
    opts?: InitOptions,
```

Comment (tomayac) [reply index.d.ts]: Looks like. Added.

**Comment by proof-llc** [comment]
Shouldn't SqliteClient be included?

**Comment by tomayac** [comment]
This is based on the latest version by @jbaiter from https://github.com/sqlite/sqlite-wasm/issues/1#issuecomment-1699769282.

**Pull request closed without merging**